lib/secure_sockets/portable/freertos_plus_tcp/aws_secure_sockets.c

```diff
@@ -531,52 +532,56 @@ BaseType_t SOCKETS_Init( void )
  * and FreeRTOS+TCP threads. Assume that two or more threads may race to be the
  * first to initialize the static and handle that case accordingly.
  */
-static CK_RV prvSocketsGetCryptoSession( CK_SESSION_HANDLE * pxSession,
+static CK_RV prvSocketsGetCryptoSession( SemaphoreHandle_t *pxSessionLock,
+                                         CK_SESSION_HANDLE * pxSession,
                                          CK_FUNCTION_LIST_PTR_PTR ppxFunctionList )
 {
     CK_RV xResult = CKR_OK;
     CK_C_GetFunctionList pxCkGetFunctionList = NULL;
     static CK_SESSION_HANDLE xPkcs11Session = 0;
     static CK_FUNCTION_LIST_PTR pxPkcs11FunctionList = NULL;
-    CK_BBOOL xNeedInit = CK_FALSE;
-    CK_SESSION_HANDLE xTempPkcs11Session = 0;
-    CK_FUNCTION_LIST_PTR pxTempPkcs11FunctionList = NULL;
+    static StaticSemaphore_t xStaticSemaphore;
+    static SemaphoreHandle_t xSessionLock = NULL;
     CK_ULONG ulCount = 0;
     CK_SLOT_ID *pxSlotIds = NULL;
 
-    /* Check if one-time initialization is needed.*/
+    /* Check if one-time initialization of the lock is needed.*/
     portENTER_CRITICAL();
-    if( 0 == xPkcs11Session )
+    if( NULL == xSessionLock )
     {
-        xNeedInit = CK_TRUE;
+        xSessionLock = xSemaphoreCreateMutexStatic( &xStaticSemaphore );
     }
+
+    *pxSessionLock = xSessionLock;
     portEXIT_CRITICAL();
 
-    if( CK_TRUE == xNeedInit )
+    /* Check if one-time initialization of the crypto handle is needed.*/
+    xSemaphoreTake( xSessionLock, portMAX_DELAY );
+    if( 0 == xPkcs11Session )
     {
         /* One-time initialization. */
 
         /* Ensure that the PKCS#11 module is initialized. We don't keep the
-        above lock here, since we don't want to make assumptions about hardware
+        scheduler stopped here, since we don't want to make assumptions about hardware
         requirements for accessing a crypto module. */
 
-        if( 0 == xResult )
+        if( CKR_OK == xResult )
```

Comment:
NP: Technically we don't need this first check if CKR_OK == xResult, but I don't think it hurts anything.

Reply:
Fixed.